### PR TITLE
Addresses #1497: FeatureLocation invalid input exception

### PR DESCRIPTION
**Pull request by JamesJeffryes**
FeatureLocation will now raise a value error if supplied an start location greater than the specified end location

This pull request addresses issue #1497 

I hereby agree to dual licence this and any previous contributions under both
the _Biopython License Agreement_ **AND** the _BSD 3-Clause License_.

I have read the ``CONTRIBUTING.rst`` file and understand that AppVeyor and
TravisCI will be used to confirm the Biopython unit tests and ``flake8`` style
checks pass with these changes.

I am happy be thanked by name in the ``NEWS.rst`` and ``CONTRIB.rst`` files,
and have added myself to those files as part of this pull request. (*This
acknowledgement is optional. Note we list the names sorted alphabetically.*)

### CODE REVIEW

**Comment by JamesJeffryes** [comment]
When testing I discovered my change must support a feature location with an UnknownPosition like FeatureLocation(42, UnknownPosition()). However the __len__ of a feature location with an UnknownPosition is invalid. 
```
Traceback (most recent call last):
  File "/Users/jjeffryes/anaconda3/lib/python3.6/site-packages/IPython/core/interactiveshell.py", line 2881, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-22-d5f7e1d31af9>", line 1, in <module>
    len(FeatureLocation(UnknownPosition(), 42))
  File "/Users/jjeffryes/anaconda3/lib/python3.6/site-packages/Bio/SeqFeature.py", line 787, in __len__
    return int(self._end) - int(self._start)
TypeError: int() argument must be a string, a bytes-like object or a number, not 'UnknownPosition'
```
I think it's right to have an exception raised in this case but would you like me to add a more informative error message to this PR?

**Comment by peterjc** [comment]
Good point - I'd forgotten about the annoying special case, yes, if either start or end is ``UnknownPosition`` the order check will fail.

As to the exception with ``__len__`` in this case, I think the current ``TypeError`` exception is reasonably clear as is - these locations are fairly rare so I wouldn't worry too much about making that exception more informative.

(Note now that I've merged #1500 the NumPy doctest failures on the master branch should be solved #1496; you were probably ignoring them anyway)

File: Bio/SeqFeature.py [hunk 1]
```diff
@@ -669,10 +669,10 @@ def __init__(self, start, end, strand=None, ref=None, ref_db=None):
             self._end = ExactPosition(end)
         else:
             raise TypeError("end=%r %s" % (end, type(end)))
-        if self._start > self._end:
+        if self.start.position and self.end.position and self.start > self.end:
```
Comment (peterjc) [inline Bio/SeqFeature.py]: Sadly zero is a valid start or end position, so I think this would miss some corner cases

File: Tests/test_SeqFeature.py [hunk 2]
```diff
@@ -8,7 +8,8 @@
 import unittest
 from os import path
 from Bio import SeqIO
-from Bio.SeqFeature import FeatureLocation, AfterPosition, BeforePosition, CompoundLocation
+from Bio.SeqFeature import FeatureLocation, AfterPosition, BeforePosition, \
+    CompoundLocation, UnknownPosition
```
Comment (peterjc) [inline Tests/test_SeqFeature.py]: Could you use two import lines, rather than the slash continuation here.

File: Tests/test_SeqFeature.py [hunk 3]
```diff
@@ -84,6 +85,10 @@ def test_start_before_end(self):
         with self.assertRaisesRegex(ValueError, err):
             FeatureLocation(BeforePosition(42), AfterPosition(23), -1)
 
+        # Features with UnknownPositions should pass test
+        FeatureLocation(42, UnknownPosition())
+        FeatureLocation(UnknownPosition(), 42)
+
```
Comment (peterjc) [inline Tests/test_SeqFeature.py]: Include some zero positions, and also special case where start and end are equal (e.g. location of digest enzyme cutting site between two bases)?

**Comment by JamesJeffryes** [comment]
So the travis build seems to have failed on a network error. I can't make much sense of why one of the appveyor builds failed it's tests

**Comment by peterjc** [comment]
As you say, the tests look to have failed due to a network issue. Sadly they would have failed anyway on AppVeyor due to #1496 which I am working on now, #1502

**Comment by peterjc** [comment]
@bow - does this seem fine to you too?

**Comment by JamesJeffryes** [comment]
@peterjc Any further feedback on this?

**Comment by peterjc** [comment]
I'm happy. @bow if this looks OK to you, either of us could go ahead and rebase-merge this.

**Comment by peterjc** [comment]
Thanks Bow, merging. Thank you James!
